### PR TITLE
refactor: convert action to use TS

### PR DESCRIPTION
**Pull request by michael-siek**
Closes: https://github.com/dequelabs/axe-linter-action/issues/19
Closes: https://github.com/dequelabs/axe-linter-action/issues/47
Closes: https://github.com/dequelabs/axe-linter-action/issues/34
Closes: https://github.com/dequelabs/axe-linter-action/issues/26
Closes: https://github.com/dequelabs/axe-linter-action/issues/32

qa notes: run a full regression

### CODE REVIEW

**Comment by stephenmathieson** [comment]
👍 for adding tests, but this is quite a bit more code to maintain. Rather than a full rewrite, have you considered using a bash testing framework to achieve test coverage?

File: action.yml [hunk 1]
```diff
@@ -1,8 +1,8 @@
 name: Run axe Accessibility Linter
```
Comment (michael-siek) [inline action.yml:1]: moving over to typescript removes the need to use third party actions.

File: action.yml [hunk 2]
```diff
@@ -15,30 +15,5 @@ inputs:
...
-    - name: Run axe linter
-      if: steps.changed_files.outputs.any_changed == 'true'
-      run: ${{ github.action_path }}/axe-linter.sh
-      env:
-        GITHUB_TOKEN: ${{ inputs.github_token }}
-        API_KEY: ${{ inputs.api_key }}
-        CHANGED_FILES: ${{ steps.changed_files.outputs.all_changed_files }}
-        AXE_LINTER_URL: ${{ inputs.axe_linter_url }}
-      shell: bash
+  using: 'node20'
```
Comment (michael-siek) [inline action.yml:18]: https://github.com/dequelabs/axe-linter-action/issues/59 - tech debt ticket to upgrade to node22 and remove `node-fetch`

File: src/run.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    if (errorCount > 0) {
+      core.setFailed(
+        `Found ${errorCount} accessibility issue${errorCount === 1 ? '' : 's'}`
+      )
+    }
+  } catch (error) {
+    if (error instanceof Error) {
+      core.setFailed(error.message)
+    } else {
+      core.setFailed('An unexpected error occurred')
```
Comment (zlayaAvocado) [inline src/run.ts:56]: do you think there is a way to make the error more helpful than just unexpected? :D

File: src/linter.test.ts [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        await lintFiles(files, apiKey, axeLinterUrl, linterConfig)
+        assert.fail('Should have thrown an error')
+      } catch (error) {
+        // Verify that the caught error is our non-Error object
+        assert.isFalse(
+          error instanceof Error,
+          'Error should not be an Error instance'
+        )
+        assert.deepEqual(
+          error as any,
```
Comment (zlayaAvocado) [inline src/linter.test.ts:315]:
```suggestion
          error,
```

File: src/linter.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+      if (result.error) {
+        throw new Error(result.error)
+      }
+
+      const errors = result.report.errors
+      totalErrors += errors.length
+
+      // Report errors using GitHub annotations
+      // There is a limit of 10 warning annotations and 10 error annotations per step
```
Comment (zlayaAvocado) [inline src/linter.ts:47]: can we make a note about it somewhere else (e.g. README?) since this can be confusing otherwise

**Review by zlayaAvocado — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      - main
+
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-node@v4
+        with:
+          node-version: 20
```
Comment (stephenmathieson) [inline .github/workflows/tests.yml:16]: Rather than creating a tech debt ticket, why don't we just use Node.js v22 now?

```suggestion
          node-version: 22
```

File: .husky/pre-commit [hunk 7]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+cd "$(dirname -- "$0")/../."
+yarn build
```
Comment (stephenmathieson) [inline .husky/pre-commit:4]: I am worried that running `yarn build` during the commit process will interfere with fixing bugs. Do we really need this? What are we solving by running the build here?

File: .husky/pre-commit [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+
+git add .
+
+npx lint-staged
```
Comment (stephenmathieson) [inline .husky/pre-commit:8]: IIRC husky runs these scripts (`./.husky/*`) with `node_modules/.bin` in its pseudo-$PATH, so `npx` is not needed here.

```suggestion
lint-staged
```

File: .nvmrc [hunk 9]
```diff
@@ -0,0 +1 @@
+20
```
Comment (stephenmathieson) [inline .nvmrc:1]:
```suggestion
22
```

File: action.yml (hunk 2 above)
Comment (stephenmathieson) [reply action.yml:18]:
```suggestion
  using: 'node22'
```

File: package.json [hunk 10]
```diff
@@ -0,0 +1,45 @@
+{
+  "name": "axe-linter-action",
```
Comment (stephenmathieson) [inline package.json:2]: Just to be safe:

```suggestion
  "name": "@deque/axe-linter-action",
```

File: package.json [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  "version": "1.0.0",
```
Comment (stephenmathieson) [inline package.json:3]: We should prevent a developer from accidentally publishing this:

```suggestion
  "version": "1.0.0",
  "private": true,
```

File: package.json [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  "scripts": {
+    "prebuild": "rimraf dist",
+    "build": "ncc build src/index.ts --license licenses.txt",
+    "lint": "eslint .",
+    "fmt": "prettier --write \"src/**/*.{js,ts,tsx,json,md,css,html}\"",
```
Comment (stephenmathieson) [inline package.json:8]: No need to manually specify a glob, Prettier will do it on its own:

```suggestion
    "fmt": "prettier --write .",
```

File: package.json [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    "prepare": "husky",
+    "test": "mocha 'src/**/*.test.ts'",
```
Comment (stephenmathieson) [inline package.json:10]: Just curious - why Mocha rather than `node:test`? I've been using the stdlib framework for the past few months (examples: [Launchpad](https://github.com/dequelabs/launchpad/blob/develop/packages/ui-issues/src/configuration.test.tsx#L1), [Trombone](https://github.com/dequelabs/trombone/blob/develop/api-client-js/src/Client.test.ts#L2)) and am very happy with it.

File: src/git.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+      owner: context.repo.owner,
+      repo: context.repo.repo,
+      base,
+      head
+    })
+
+    return (
+      response.data.files
+        ?.filter((file) =>
+          FILE_PATTERNS.some((pattern) => minimatch(file.filename, pattern))
```
Comment (stephenmathieson) [inline src/git.ts:36]: Does `minimatch` do case insensitive search? Will `readme.MarKdoWN` be matched here?

File: src/linter.test.ts [hunk 15]
```diff
@@ -0,0 +1,337 @@
+import { assert } from 'chai'
+import * as sinon from 'sinon'
+import * as core from '@actions/core'
+import nock from 'nock'
+import { lintFiles } from './linter'
+import type { LinterResponse } from './types'
+
+type MockResponses = {
+  [key: string]: LinterResponse
+}
```
Comment (stephenmathieson) [inline src/linter.test.ts:10]: nit-

```suggestion
type MockResponses = Record<string, LinterResponse>
```

File: src/linter.ts [hunk 16]
```diff
@@ -0,0 +1,69 @@
+import * as core from '@actions/core'
+import { readFile } from 'fs/promises'
+import type { LinterResponse } from './types'
+import fetch from 'node-fetch'
+
+export async function lintFiles(
+  files: string[],
+  apiKey: string,
+  axeLinterUrl: string,
+  linterConfig: Record<string, unknown>
```
Comment (stephenmathieson) [inline src/linter.ts:10]: Why is this `unknown`? Shouldn't we know the config type here?

File: src/linter.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+): Promise<number> {
+  let totalErrors = 0
+
+  for (const file of files) {
+    try {
+      const fileContents = await readFile(file, 'utf8')
```
Comment (stephenmathieson) [inline src/linter.ts:16]: Is there any benefit to this being asynchronous (rather than `readFileSync()`)?

File: src/linter.ts [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+          Authorization: apiKey
+        },
+        body: JSON.stringify({
+          source: fileContents,
+          filename: file,
+          config: linterConfig
+        })
+      })
+
+      const result = (await response.json()) as LinterResponse
```
Comment (stephenmathieson) [inline src/linter.ts:37]: We need to handle network errors and unhealthy server responses.

```suggestion
      const response = await fetch(`${axeLinterUrl}/lint-source`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: apiKey
        },
        body: JSON.stringify({
          source: fileContents,
          filename: file,
          config: linterConfig
        })
      })

      if (!response.ok) {
        // ...
      }
      
      var contentType = response.headers.get("Content-Type")
      if (!contentType || !contentType.includes("json")) {
        // ...
      }

      const result = (await response.json()) as LinterResponse
```

File: src/linter.ts [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+          endColumn: error.endColumn,
+          title: 'Axe Linter'
+        })
+      }
+    } catch (error) {
+      if (error instanceof Error) {
+        throw new Error(`Error processing ${file}: ${error.message}`)
+      }
+      throw error
+    }
```
Comment (stephenmathieson) [inline src/linter.ts:64]: This `catch` block seems unnecessary. Can we let errors bubble rather than re-throwing them here?

File: src/run.ts [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    // Remove trailing slash if present
+    inputs.axeLinterUrl = inputs.axeLinterUrl.replace(/\/$/, '')
+
+    // Get changed files
+    const changedFiles = await getChangedFiles(inputs.githubToken)
+
+    if (changedFiles.length === 0) {
+      core.info('No files to lint')
```
Comment (stephenmathieson) [inline src/run.ts:23]: nit - the only reason a user would want to see this log is if they're debugging an issue.

```suggestion
      core.debug('No files to lint')
```

File: src/run.ts [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+
+    // Load linter config if exists
+    let linterConfig = {}
+    try {
+      const configFile = await readFile('axe-linter.yml', 'utf8')
+      const parsedConfig = load(configFile)
+      if (parsedConfig && typeof parsedConfig === 'object') {
+        linterConfig = parsedConfig
+      }
+    } catch {
```
Comment (stephenmathieson) [inline src/run.ts:35]: If there is an error, we should log it.
```suggestion
    } catch (error) {
      core.error("An error happened and stuff", error)
```

**Review by stephenmathieson — DISMISSED** [review]

File: action.yml [hunk 22]
```diff
@@ -15,30 +15,5 @@ inputs:
...
+  main: 'dist/index.js'
```
Comment (stephenmathieson) [inline action.yml:19]: Rather than keeping compiled code in Git, we could follow `action-sync-branches` and use Docker 🤷

File: .github/workflows/tests.yml (hunk 6 above)
Comment (michael-siek) [reply .github/workflows/tests.yml:16]: The actual runner doesn't support node 22 which is why I used node 20 here to match the node runner version

File: .husky/pre-commit (hunk 7 above)
Comment (michael-siek) [reply .husky/pre-commit:4]: Making sure that the dist build is always up to date this is a similar structure to other deque actions

File: .github/workflows/tests.yml (hunk 6 above)
Comment (stephenmathieson) [reply .github/workflows/tests.yml:16]: Yes it does. See https://github.com/dequelabs/launchpad/blob/develop/.github/actions/dependencies/action.yml#L11-L13

File: src/linter.ts (hunk 16 above)
Comment (stephenmathieson) [reply src/linter.ts:11]: If we're not going to strongly type our code, why use TypeScript?

File: src/git.test.ts
Comment (zlayaAvocado) [inline src/git.test.ts:1]: nice - 1000 edge-cases :D

File: src/utils.ts
Comment (zlayaAvocado) [inline src/utils.ts:1]: this wouldn't cover some exceptions that don't end at `s` but it's fine, if we need we can improve it later or use an existing package

File: .github/workflows/tests.yml (hunk 6 above)
Comment (dbjorge) [reply .github/workflows/tests.yml:16]: @stephenmathieson The `setup-node` action does, but that's not what controls the actual runtime environment, that just controls what we're running our build/tests on. The thing that controls what node version a user will actually be using to *run* the action code is the `runs: { using: 'node20' }` directive in `action.yml`, where `node20` is the only supported option. https://github.com/actions/runner/issues/3600 tracks support for custom actions using `node22` (but I would be surprised if GitHub actually implemented that, historically they've supported every-other-LTS at best)

File: src/linter.ts (hunk 16 above)
Comment (dbjorge) [reply src/linter.ts:11]: I think it's reasonable/good for the action to treat the config as opaque/unknown **iff the following are all true**:

- The action doesn't need to actually use the config (if it needs just parts, I'm fine with typing/validating just those parts)
- The service API it calls is responsible for parsing/validating the config
- When the service API detects an invalid config, it emits an actionable error response which the action displays reasonably to the user

**Review by zlayaAvocado — APPROVED** [review]
my comments are responded

File: src/linter.ts (hunk 16 above)
Comment (WilcoFiers) [reply src/linter.ts:11]: I'm with Michael on this. This code doesn't use the types at all. If it was easy to get the types in I'd say just do it anyway, but there isn't. We don't have the types published publicly. We'd either have to maintain a duplicate of a type that we don't use, or we'd have to add a publication step for an unused type. There's no value in that.

File: action.yml (hunk 22 above)
Comment (michael-siek) [reply action.yml:19]: I spoke with the team and this could be considered a breaking change
